### PR TITLE
tests(clustering/rpc): rpc mock/hook

### PR DESCRIPTION
**Pull request by StarlightIbuki**
### Summary

RPC needs mock tests

doc: https://github.com/Kong/kong/blob/rpc-mock/spec/helpers/rpc_mock/readme.md

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)

### Issue reference

Fix KAG-5551

### CODE REVIEW

File: spec/01-unit/19-hybrid/04-rpc_spec.lua [hunk 1]
```diff
@@ -0,0 +1,68 @@
+-- by importing helpers, we initialize the kong PDK module
+local helpers = require "spec.helpers"
+local server = require("spec.helpers.rpc_mock.server")
+local client = require("spec.helpers.rpc_mock.client")
+
+describe("rpc v2", function()
```
Comment (chronolaw) [inline spec/01-unit/19-hybrid/04-rpc_spec.lua:6]: The title should be `sync v2` or `rpc sync`

File: spec/helpers/rpc_mock/server.lua [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+function _M.start(self)
+  local bp, db = helpers.get_db_utils(strategy, nil, { "rpc-proxy" })
+
+  local plugin = db.plugins:insert({
+    name = "rpc-proxy",
+    config = {},
+  })
+
+  assert(helpers.start_kong(self))
```
Comment (chobits) [inline spec/helpers/rpc_mock/server.lua:41]: start a cp to proxy client's rpc request to DP

File: spec/helpers/rpc_mock/server.lua [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+  self.proxy_client = client.new({
```
Comment (chobits) [inline spec/helpers/rpc_mock/server.lua:43]: client -> mock cp --> tested DP

File: spec/helpers/rpc_mock/server.lua [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    cluster_control_plane = self.cluster_listen,
+  })
+
+  if self.interception then
+    self:enable_inception()
+  end
+
+  self.proxy_client.callbacks:register("kong.rpc.proxy.mock", function(proxy_id, proxy_payload)
```
Comment (chobits) [inline spec/helpers/rpc_mock/server.lua:51]: mock cp will rpc call this callback to forward `[the RPC call request from tested DP]` to `[client]`

File: spec/01-unit/19-hybrid/04-rpc_spec.lua [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  describe("full sync pagination", function()
+    describe("server side", function()
+      local server_mock
+      lazy_setup(function()
+        server_mock = server.new()
+        assert(server_mock:start())
+
+        assert(helpers.start_kong({
```
Comment (chobits) [inline spec/01-unit/19-hybrid/04-rpc_spec.lua:14]: tested DP

File: spec/helpers/rpc_mock/client.lua [hunk 6]
```diff
@@ -0,0 +1,58 @@
+-- by importing helpers, we ensure the kong PDK module is initialized
+local helpers = require "spec.helpers"
+local rpc_mgr = require("kong.clustering.rpc.manager")
+local default_cert = require("spec.helpers.rpc_mock.default").default_cert
+
```
Comment (chronolaw) [inline spec/helpers/rpc_mock/client.lua:5]: style: two blank lines between blocks.

File: spec/helpers/rpc_mock/server.lua [hunk 7]
```diff
@@ -0,0 +1,199 @@
+local helpers = require("spec.helpers")
+local client = require("spec.helpers.rpc_mock.client")
+local default_cert = require("spec.helpers.rpc_mock.default").default_cert
+
```
Comment (chobits) [inline spec/helpers/rpc_mock/server.lua:4]: I strongly recommend you to write a callback and diagram in the comment like

```
diagram:

mock client -> mock server -> tested DP

callback
func1
 -> func2
  -> func3 ...

```

otherwise other developers or reviewers find it hard to understand, today i talked with @chronolaw , he found it hard to understand what the server.lua does here.

Comment (chobits) [reply spec/helpers/rpc_mock/server.lua:4]: a good way is that you could write a test case to run this `mock client -> mock server -> tested DP` progress, i find there is no case for `mock server<>tested DP`. Then we could explain the progress with that case, that might be easy for reviewers.

File: spec/internal/misc.lua [hunk 8]
```diff
@@ -287,6 +289,49 @@ local function use_old_plugin(name)
...
+local function patch_worker_events()
+  if not kong then
+    return
+  end
+  
+  if kong.worker_events then
+    return
+  end
+
+  kong.worker_events = require "resty.events.compat"
```
Comment (chronolaw) [inline spec/internal/misc.lua:327]:
```suggestion
  kong.worker_events = require("resty.events.compat")
```

File: spec/01-unit/09-balancer/03-consistent_hashing_spec.lua [hunk 9]
```diff
@@ -319,10 +319,10 @@ describe("[consistent_hashing]" .. (enable_new_dns_client and "[new dns]" or "")
         res[addr..":"..port] = (res[addr..":"..port] or 0) + 1
         res[host..":"..port] = (res[host..":"..port] or 0) + 1
       end
-      assert(15 == res["1.2.3.4:123"] or nil == res["1.2.3.4:123"], "mismatch")
-      assert(15 == res["mashape.com:123"] or nil == res["mashape.com:123"], "mismatch")
-      assert(15 == res["5.6.7.8:321"] or nil == res["5.6.7.8:321"], "mismatch")
-      assert(15 == res["getkong.org:321"] or nil == res["getkong.org:321"], "mismatch")
+      assert.is_true(15 == res["1.2.3.4:123"] or nil == res["1.2.3.4:123"], "mismatch")
```
Comment (chronolaw) [inline spec/01-unit/09-balancer/03-consistent_hashing_spec.lua:322]: Is this change related to rpc mock? If not  I suggest to move them to another PR.

File: spec/02-integration/01-helpers/05-rpc-mock_spec.lua [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+      client_mock:start()
+      client_mock:wait_until_connected()   
+    end)
+
+    lazy_teardown(function()
+      helpers.stop_kong(nil, true)
+      client_mock:stop()
+    end)
+
+    it("client->CP", function()
```
Comment (chronolaw) [inline spec/02-integration/01-helpers/05-rpc-mock_spec.lua:125]: Should be `DP -> CP` or `client -> server`?

File: spec/02-integration/01-helpers/05-rpc-mock_spec.lua [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+      local res, err = client_mock:call("control_plane", "kong.sync.v2.get_delta", { default = { version = 0,},})
+      assert.is_nil(err)
+      assert.is_table(res and res.default and res.default.deltas)
+
+      local res, err = client_mock:call("control_plane", "kong.sync.v2.unknown", { default = { },})
+      assert.is_string(err)
+      assert.is_nil(res)
+    end)
+
+    it("CP->client", function()
```
Comment (chronolaw) [inline spec/02-integration/01-helpers/05-rpc-mock_spec.lua:135]: Should be `CP -> DP` or `server -> client`?

File: spec/fixtures/custom_plugins/kong/plugins/rpc-debug/handler.lua [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+local _M = {
+  PRIORITY = 1000,
+  VERSION = kong_meta.version,
+}
+
+local original_callbacks = {}
+local inc_id = 0
+
+function _M.init_worker()
+  kong.rpc.callbacks:register("kong.rpc.debug.register", function(node_id, register_payload)
```
Comment (chronolaw) [inline spec/fixtures/custom_plugins/kong/plugins/rpc-debug/handler.lua:12]: Should we rename it to `rpc-testing`? It dose not look like `debug`.

File: spec/helpers.lua [hunk 13]
```diff
@@ -226,4 +226,10 @@ local wait = reload_module("spec.internal.wait")
   make_temp_dir = misc.make_temp_dir,
 
   build_go_plugins = cmd.build_go_plugins,
+
+  get_node_id = misc.get_node_id,
```
Comment (chronolaw) [inline spec/helpers.lua:230]: Perhaps we should not export these apis in `helpers`, could we use `misc` directly?

File: spec/helpers/rpc_mock/client.lua [hunk 14]
```diff
@@ -0,0 +1,68 @@
+-- by importing helpers, we ensure the kong PDK module is initialized
+local helpers = require "spec.helpers"
+local rpc_mgr = require("kong.clustering.rpc.manager")
+local default_cert = require("spec.helpers.rpc_mock.default").default_cert
+local uuid = require "kong.tools.uuid"
```
Comment (chronolaw) [inline spec/helpers/rpc_mock/client.lua:5]:
```suggestion
local uuid = require("kong.tools.uuid")
```

File: spec/helpers/rpc_mock/client.lua [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+
+local _M = {}
+
+
+local default_dp_conf = {
+  role = "data_plane",
+  cluster_control_plane = "localhost:8005",
+}
+
+setmetatable(default_dp_conf, { __index = default_cert })
```
Comment (chronolaw) [inline spec/helpers/rpc_mock/client.lua:16]:
```suggestion
setmetatable(default_dp_conf, { __index = default_cert, })
```

File: spec/helpers/rpc_mock/client.lua [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+  for _, socket in pairs(rpc_mgr.clients) do
+    for conn in pairs(socket) do
+      pcall(conn.stop, conn)
+    end
+  end
+end
+
+
+local function client_is_connected(rpc_mgr)
+  for _, socket in pairs(rpc_mgr.clients) do
```
Comment (chronolaw) [inline spec/helpers/rpc_mock/client.lua:37]: Should we use `table.isempty`?

File: spec/02-integration/01-helpers/05-rpc-mock_spec.lua (hunk 10 above)
Comment (StarlightIbuki) [reply spec/02-integration/01-helpers/05-rpc-mock_spec.lua:125]: They are not the same. `client` here is a mocked client, while CP is a real CP

File: spec/01-unit/09-balancer/03-consistent_hashing_spec.lua (hunk 9 above)
Comment (StarlightIbuki) [reply spec/01-unit/09-balancer/03-consistent_hashing_spec.lua:322]: @chronolaw This is a source of flaky and it fails the CI, making this PR hard to merge

File: spec/02-integration/01-helpers/05-rpc-mock_spec.lua (hunk 11 above)
Comment (StarlightIbuki) [reply spec/02-integration/01-helpers/05-rpc-mock_spec.lua:135]: Same here

File: spec/fixtures/custom_plugins/kong/plugins/rpc-debug/handler.lua (hunk 12 above)
Comment (StarlightIbuki) [reply spec/fixtures/custom_plugins/kong/plugins/rpc-debug/handler.lua:12]: I can be used for debug purpose and the connected client act as a debugger

File: spec/helpers.lua (hunk 13 above)
Comment (StarlightIbuki) [reply spec/helpers.lua:230]: I suppose `get_node_id` has more general usages

File: spec/helpers/rpc_mock/client.lua (hunk 16 above)
Comment (StarlightIbuki) [reply spec/helpers/rpc_mock/client.lua:37]: I don't think it's performance critical for a test function

Comment (chronolaw) [reply spec/helpers/rpc_mock/client.lua:37]: `table.isempty` is easy to understand for reviewers.

File: spec/02-integration/01-helpers/05-rpc-mock_spec.lua [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+
+      local dp_id = get_node_id("servroot")
+
+      server_mock:wait_for_node(dp_id)
+
+      assert(server_mock:call(dp_id, "kong.sync.v2.notify_new_version", { default = { new_version = 100, } }))
+
+      -- the mock should have been called
+      helpers.wait_until(function()
+        return client_version
```
Comment (chronolaw) [inline spec/02-integration/01-helpers/05-rpc-mock_spec.lua:89]: Should we add `assert(client_version == 100)`?

File: spec/fixtures/custom_plugins/kong/plugins/rpc-debug/handler.lua [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+
+          if res.posthook then
+            res, err = kong.rpc:call(node_id, "kong.rpc.debug.posthook", { call_id = id, method = proxy_api, node_id = client_id, payload = {result = origin_res, error = origin_err} })
+            if not res then
+              return nil, "Failed to call post hook(" .. node_id .. "): " .. err
+            end
+
+            return res.result, res.error
+          end
+        elseif res.mock then
```
Comment (chronolaw) [inline spec/fixtures/custom_plugins/kong/plugins/rpc-debug/handler.lua:54]: blank line before `elseif`

File: spec/helpers.lua (hunk 13 above)
Comment (chronolaw) [reply spec/helpers.lua:230]: Currently they are only used in rpc mock, we could move it to misc now and expose them in the future.

File: spec/02-integration/01-helpers/05-rpc-mock_spec.lua (hunk 17 above)
Comment (StarlightIbuki) [reply spec/02-integration/01-helpers/05-rpc-mock_spec.lua:89]: No this is not how it should work

File: spec/helpers/rpc_mock/client.lua (hunk 16 above)
Comment (StarlightIbuki) [reply spec/helpers/rpc_mock/client.lua:37]: The main concern here is to reduce dependency. Anyway let's do it

File: spec/02-integration/01-helpers/05-rpc-mock_spec.lua [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+      }))
+    end)
+
+    lazy_teardown(function()
+      mocked_cp:stop(true)
+      helpers.stop_kong(nil, true)
+    end)
+
+    it("interception", function()
+      -- this will eventually cause DP to make a "kong.sync.v2.get_delta" call to the CP
```
Comment (chobits) [inline spec/02-integration/01-helpers/05-rpc-mock_spec.lua:68]:
```suggestion
      -- POST /service entity to CP admin API
      -- this will eventually cause DP to make a "kong.sync.v2.get_delta" call to the CP
```

File: spec/02-integration/01-helpers/05-rpc-mock_spec.lua [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+        return { default = { version = 100, deltas = {} } }
+      end)
+      -- reattach the debugger for the mock to take effect
+      mocked_cp:attach_debugger()
+
+      -- wait until the data plane node is connected
+      local dp_id = get_node_id("servroot")
+      mocked_cp:wait_for_node(dp_id)
+
+      -- make a call from the mocked cp
```
Comment (chobits) [inline spec/02-integration/01-helpers/05-rpc-mock_spec.lua:90]:
```suggestion
      -- make a call from the mocked cp:
      --    (1) mocked_cp -- rpc:call --> DP's kong.sync.v2.notify_new_version
      --    (2) DP -- rpc:call --> kong.sync.v2.get_delta -> called = true
```

Comment (chobits) [reply spec/02-integration/01-helpers/05-rpc-mock_spec.lua:90]: need an example rpc service : `kong/clustering/services/example`

**Review by chobits — APPROVED** [review]

File: spec/02-integration/01-helpers/05-rpc-mock_spec.lua [hunk 21]
```diff
@@ -0,0 +1,243 @@
...
+        local proxy_client = assert(helpers.proxy_client())
+
+        body = assert.res_status(200, proxy_client:send {
+          method = "GET",
+          path = "/",
+          headers = {
+            ["x-greeting"] = "world",
+          }
+        })
+      end, 10)
```
[the remark this answers is not in the document]
Comment (StarlightIbuki) [reply spec/02-integration/01-helpers/05-rpc-mock_spec.lua]: A mindless copy & paste. Modified

File: spec/02-integration/01-helpers/05-rpc-mock_spec.lua [hunk 22]
```diff
@@ -0,0 +1,240 @@
+local helpers = require("spec.helpers")
+local misc = require("spec.internal.misc")
+local cp = require("spec.helpers.rpc_mock.cp")
+local dp = require("spec.helpers.rpc_mock.dp")
+local setup = require("spec.helpers.rpc_mock.setup")
+local get_node_id = misc.get_node_id
+local DP_PREFIX = "servroot_dp"
+
+describe("rpc mock/hooc", function()
```
Comment (chronolaw) [inline spec/02-integration/01-helpers/05-rpc-mock_spec.lua:9]: `rpc mock/hook`?

**Review by ADD-SP — APPROVED** [review]
